### ISSUE TITLE
Discuss: Break Compatibility?

### ISSUE THREAD

**Issue by AdamSpeight2008** [dotnet/roslyn#265]
When would it be beneficial / profitable to break backwards compatibly?

Let's say some future release does that, and then becomes the new base target for compatibility.
What would you add? What would you remove? What would you enhance / improve?

* No non-generic collections
* `System.Void` a first class type.
* `Action<T>` is just an alias for `Func<T,void>`

**Comment by HaloFour** [comment]
I would hope not.  This would immediately render a huge number of existing assemblies as obsolete.  A language without an ecosystem may as well not exist.

As for `System.Void` I do think that there is value in such a concept but that it can fit into the existing language and framework.  It would simply be a placeholder type, a one-byte `struct` with no public value that could be assigned, or an empty `class` with no public constructor.  This is how F# and Java do it and it works pretty well.  You don't get the duck-typing of `Action<T>` to `Func<T, Void>` but if the type was a first class citizen in the BCL nothing would stop the language from allowing conversions between them.

**Comment by sharwell** [comment]
After watching the progress of #166, I'm not sure these very broad topics will prove as useful one might hope. The compatibility issue is fundamental to the discussion of the merits of the various specific examples you gave, and will naturally appear in those topics with both merits and disadvantages.

**Comment by AdamSpeight2008** [comment]
@HaloFour A lot of the libraries are being re-implemented / modified to be portable libraries. 

On `Action<T>` it would be there at all, just `Func<T>`, which would in turn make the BCL smaller.

Imagine if .net had generics from the beginning, how small the BCL could have been.

**Comment by AdamSpeight2008** [comment]
VB.net is still largely compatible with VB6 (and earlier versions) Should the training wheels and ties finally be removed? Or made optional (incompatible by default)?

**Comment by HaloFour** [comment]
For the most part those libraries still maintain compatibility with aspects of the run time going out of the way to make that possible, such as with type-forwarding.  Your proposals would have wide sweeping effects on a large percentage of the underlying run time as well as a large body of existing work.

Bringing VB isn't the conversation is irrelevant as .NET has absolutely no ties to the previous VB VM and run time.  And you've been able to remove the dependency on Microsoft.VisualBasic.dll for quite some time (undocumented initially, IIRC).

**Comment by AdamSpeight2008** [comment]
VB6 language constructs still exist.

**Comment by mirhagk** [comment]
I think it's worth considering this perhaps, but only if it's really worthwhile. Python is still split between versions for it's minor breaking changes.

What would make it worthwhile? Non-Nullable types would be one thing, but it'd need a lot more than that. If you're going to break something, do it as infrequently as possible, and make it worthwhile.

**Comment by HaloFour** [comment]
That they do, but that's not relevant because they were re-implementations and we're not talking about breaking the CLR/BCL in order to excise them.

I think that it's a very good thing that .NET 1.0 assemblies still work perfectly fine in .NET 4.5.2.  There is the question of targeting, such as with portable libraries, but that was the case even in the 1.x time frame what with the Compact Framework.  But you could still take a Compact Framework assembly from 2002 and run it on the full .NET Framework today and it would work fine.

**Comment by AdamSpeight2008** [comment]
+ Override-able defaults instances. 
One implementation would involve modifying `Object` which would break everyone.

**Comment by SolalPirelli** [comment]
This kind of discussion should be on Reddit, or some other forum-like website that allows threaded discussions, IMHO. It's not very useful here, apart from exploding the notifications counter of the C# team, who *might* have a few more important things to do. :)

**Comment by AdamSpeight2008** [comment]
@SolalPirelli Which why I preferred the codeplex setup. Discussions != Issues

**Comment by jwooley** [comment]
One issue with merging Func<T> with Action<> is VB treats these very differently. For example:

Dim X As Func(Of T) = Function() a = b
Dim Y As Action(Of ) = Sub() a = b

In the case of X, a = b is an equality comparison. In the case of Y, a = b is an assignment of b to a. In order for the distinction to go away, we would need a separate assignment operator in VB which would be a much bigger breaking change.

**Comment by AdamSpeight2008** [comment]
@jwooley  I see no issue because Lambdas can target on of two possible types:- `Func<T>` or `Expression<Func<T>>`. So the type has to be specified. 
A `Sub()' is a `Func( ..., Void)` so the following

```vb
Dim X As Func(Of T) = Function() a = b ' Operator =
Dim Y As Func(Of Void) = Sub() a = b ' Assignment
```
The could be an ambiguous, an error `doesn't return void`
```vb
Dim Z As Func(Of Void) = Function() a = b ' Assignment
```
*Note:* If the return of the func is void, and the lambda is a "short form" then it is assignment.
But it can easily be fixed with the follow form.
```vb
Dim Z As Func(Of Void) = Function()
                           a = b ' Assignment
                           Return Void
                         End Function
```

**Comment by lyubomyr-shaydariv** [comment]
If I could break something, I'd break the follwing:

* Introduce r/o arrays safe to be passed elsewhere until reflection is used. CLR break, right?
* Shrink `System.Object`: implement `Equals` and `GetHashCode` coming from something like `IEquatable<T>`, `ToString` from something like `IDisplayable` -- or via traits/extension methods, whatever. I just think that the root class is too heavy. This is even more observable in Java with all of such methods like `notify`, `notifyAll` and `wait` (x3 overloads).
* Rework extension methods somehow -- they look like a sort of flood to me.

**Comment by AnthonyDGreen** [comment]
@AdamSpeight2008 
Re: VB6 training wheels

Could you be more specific?

**-ADG**

**Comment by AdamSpeight2008** [comment]
@AnthonyDGreen Admittedly it maybe better suited towards Analysers. 
```vb
value = IIF( , ,)
MID
LEFT
RIGHT
ASC
CHR
RANDOMISE
```
Coding pattern
```vb
Dim x As Integer
x = 10 ' on the very next line
Dim s As String
s = "My String" ' on the very next line
```
You just know it when you see it, this style is more vB6 then VB.net.

**Comment by jwooley** [comment]
This might be a better case for a code fix than breaking change in the compiler itself. I’ve added a similar proposal on CodeCracker (https://github.com/code-cracker/code-cracker/issues/265).






Sent from Windows Mail





From: Adam Speight
Sent: ‎Saturday‎, ‎February‎ ‎7‎, ‎2015 ‎10‎:‎54‎ ‎AM
To: dotnet/roslyn
Cc: Jim Wooley





@AnthonyDGreen Admittedly it maybe better suited towards Analysers. 

value = IIF( , ,)
MID
LEFT
RIGHT
ASC
CHR
RANDOMISE

Coding pattern

Dim x As Integer
x = 10 ' on the very next line
Dim s As String
s = "My String" ' on the very next line

You just know it when you see it, this style is more vB6 then VB.net.

**Comment by theoy** [comment]
There are many great points on this thread!

I'd like to think there are times where we consider accepting different types of incompatibilities, but they usually come with really honest extremely compelling cases.

For example, with .NET generics - there was a new metadata format. While the runtime loaded older versions of IL, the lingua franca changed and almost everything that took in collections/arrays had to be updated.

I'd like to think that most new versions have this sort of sideways compatibility. While the old paradigm may still function, it may become a 2nd-class citizen in this new world. I know that after we switched to Task-based asynchrony with .NET 4.5/C# 5, that I started wrapping everything I wrote to be Task-based :smile:

Usually to consider compatibility breaks, I advocate people to look for two opportunities:
1.  Breaks that are necessary for something tremendously valuable. These are the best kind (if any), because users can easily see the value behind the break and are more likely to accept them.
2.  Breaks that can ride sidecar to other (bigger)breaks. These are not as great as the first, but sometimes if people have to deal with a big break, one more small one at the same time is somewhat palatable.

Feel free to send me any issues you find with Roslyn incompatibilities, I organize our in-house compatibility reviews and facilitate breaking change approvals :smile:
